### ISSUE TITLE
Feature request: Allow conditional compilation of (at least) lexer mappings via #[cfg()]

### ISSUE THREAD

**Issue by kleinesfilmroellchen** [lalrpop/lalrpop#681]
In [spcasm](https://github.com/kleinesfilmroellchen/spcasm), which I'm currently in the process of moving from a hand-rolled parser to LALRPOP (a super pleasant process by the way!), the hand-rolled recursive descent parser has a collection of conditionally compiled code that only parses test comments in a test build. These comments check the output of the assembler against the known good machine code and thereby serve as the test suite for the assembler. All of the code pertaining to this test output, including the TestComment token, relevant AST struct and enum fields, lexer, parser and assembler support is gated behind `#[cfg(test)]` in order to speed up production code that doesn't need it.

This poses a problem for LALRPOP, as some of the data structures change depending on compilation configuration; so it's not enough to always parse test comments and discard them if necessary. I'd have to do something like this in my lexer adapter:

```rust
extern {
	enum Token {
		"identifier" => Token::Identifier(<String>, <SourceSpan>),
                /* snip */
		"=" => Token::Equals(<SourceOffset>),
		#[cfg(test)]
		";=" => Token::TestComment(<Vec<u8>>, <SourceSpan>),
	}
}
```

For the generated code, it's probably enough to include the cfg above the respective statement in `__token_to_integer`.

For me, this would be all that is necessary. I'd be fine with having the rules for test comments always exist, just never being actioned on a regular build as the respective lexer tokens don't exist and are never emitted. After all, LALRPOP internally doesn't store my custom tokens. However, I'd love to be able to #[cfg()] other parts of the grammar, like individual rules or entire non-terminals. I recognize that that's probably a lot more complicated or even impossible, as the state machine needs to depend on the configuration in order to truly "remove" the #[cfg()]'d code as normal Rust does it. The only workaround right now is to build this logic yourself via build.rs, e.g. preprocessing the .lalrpop file or choosing between two different versions.

So TL;DR I'm asking for two things:
1. (easy) Conditionally compiled lexer mappings via #[cfg()]
2. (hard) Conditionally compiled anything (rules, nonterminals, macros?) via #[cfg()]

**Comment by Pat-Lafon** [comment]
It was relatively unknown to me but lalrpop has rudimentary cfg support in https://github.com/lalrpop/lalrpop/blob/master/doc/src/conditional-compilation.md . Not that this helps entirely here, but it does provide a road map for future improvements/help anyone who stumbles upon this issue.

**Comment by kleinesfilmroellchen** [comment]
> It was relatively unknown to me but lalrpop has rudimentary cfg support in https://github.com/lalrpop/lalrpop/blob/master/doc/src/conditional-compilation.md . Not that this helps entirely here, but it does provide a road map for future improvements/help anyone who stumbles upon this issue.

I know about this feature, but it's not helpful for point 2, which is the only way this really becomes useful in my case.

**Comment by Pat-Lafon** [comment]
@kleinesfilmroellchen This issue might be no longer useful to you but we've made some progress with cfg support in LALRPOP. Relevant to your use case is a modification I made to the lexer example here: https://github.com/lalrpop/lalrpop/blob/1b3587d46118c4798aa428dc4ae4fc9f6d752f5d/doc/lexer/src/grammar.lalrpop#L8-L29. There is also an example at https://github.com/lalrpop/lalrpop/blob/master/doc/cfg/src/cfg.lalrpop

There isn't support for cfgs everywhere in the grammar yet and there might be incompleteness we aren't aware of (compared to Rust), but it's in an early enough part of the code base pipeline(conditional compilation occurs even before macro/precedence expansion) that it can probably be extended fairly easily.

I would be interested if you or anyone else interested in this issue has feedback on these changes.

**Comment by kleinesfilmroellchen** [comment]
The current state is of course that test comments are always compiled, but I'd like to revert that (especially given that I never exposed this data to users outside of debug tooling). Will look into it!

**Comment by kleinesfilmroellchen** [comment]
@Pat-Lafon `#[cfg(test)]` is not supported, even on latest master:

```rust
error: failed to run custom build command for `spcasm v1.2.0`

Caused by:
  process didn't exit successfully:
  --- stdout
  cargo:rerun-if-changed=src/parser/asm.lalrpop
  processing file `src/parser/asm.lalrpop`
  src/parser/asm.lalrpop:66:7: 66:10 error: unexpected `cfg` argument `test`

  --- stderr
    #[cfg(test)]
          ~~~~
  thread 'main' panicked at build.rs:10:102:
  called `Result::unwrap()` on an `Err` value: Kind(InvalidData)
```

**Comment by kleinesfilmroellchen** [comment]
Additionally, using #[cfg(test)] on a token declaration does not work, since the LALRPOP codegen is (ostensibly) always run  without that attribute, so the wrong code is compiled in and it doesn't work if I change the definition of the token enum variant under test vs. non-test.
